server/app/src/main/java/com/google/heartrate/wearos/app/gatt/attributes/GattCharacteristic.java

```diff
@@ -0,0 +1,240 @@
+    private final BluetoothGattCharacteristic mBluetoothGattCharacteristic;
+
+    /** Descriptors which characteristic include. */
+    private final HashMap<UUID, GattDescriptor> descriptorByUuid = new HashMap<>();
+
+    /**
+     * Configure {@link BluetoothGattCharacteristic} with given parameters.
+     * @param uuid characteristic uuid
+     * @param properties characteristic properties
+     * @param permissions characteristic permissions
+     */
+    public GattCharacteristic(UUID uuid, int properties, int permissions) {
+        mBluetoothGattCharacteristic =
+                new BluetoothGattCharacteristic(uuid, properties, permissions);
+    }
+
+    /**
+     * Configure {@link BluetoothGattCharacteristic} with given parameters and descriptors.
+     *
+     * @param uuid characteristic uuid
+     * @param properties characteristic properties
+     * @param permissions characteristic permissions
+     * @param gattDescriptors characteristic descriptors
+     */
+    public GattCharacteristic(UUID uuid, int properties, int permissions, GattDescriptor[] gattDescriptors) {
+        this(uuid, properties, permissions);
+
+        for (GattDescriptor gattDescriptor : gattDescriptors) {
+            descriptorByUuid.put(gattDescriptor.getUUid(), gattDescriptor);
+            mBluetoothGattCharacteristic.addDescriptor(gattDescriptor.getBluetoothGattDescriptor());
+        }
+    }
+
+    /**
+     * Determine weather characteristic has descriptor with given uuid or not.
+     *
+     * @param uuid descriptor uuid to check
+     * @return true if characteristic has descriptor, false otherwise
+     */
+    public boolean hasDescriptor(UUID uuid) {
+        return descriptorByUuid.containsKey(uuid);
+    }
+
+    /**
+     * Get {@link GattDescriptor} for descriptor in characteristic with given uuid.
+     *
+     * @param descriptorUuid descriptor uuid
+     * @return {@link GattDescriptor} for characteristic
+     * @throws GattException if characteristic does not support descriptor with given uuid
+     */
+    public GattDescriptor getDescriptor(UUID descriptorUuid) throws GattException {
+        GattDescriptor descriptorManager = descriptorByUuid.get(descriptorUuid);
+        if (descriptorManager == null) {
+            throw new GattException(String.format("Characteristic %s does not have descriptor %s",
+                    getUUid(), descriptorUuid));
+        }
+        return descriptorManager;
+    }
+
+    /**
+     * Process default read characteristic request.
+     * If special actions need to be done on read request, please override this method.
+     *
+     * @param device device from which read request was received
+     * @param offset value offset
+     * @return current value of characteristic
+     * @throws GattException if value can not be read
+     */
+    public byte[] read(BluetoothDevice device, int offset) throws GattException {
+        // TODO: implement getValue with offset, not required for now
```

Comment:
In general it would be good to keep things simple as long as that is possible, so no need to wrap the full functionality of the BlueToothCharacteristic if it's never going to be used. You can always add methods later on.